tests/test_window_constraints.py

```diff
@@ -0,0 +1,534 @@
+from datetime import datetime, timedelta, timezone, time
+from pathlib import Path
+
+from unittest.mock import MagicMock, patch
+from zoneinfo import ZoneInfo
+
+from cats import main, parse_time_constraint, validate_window_constraints
+from cats.forecast import (
+    CarbonIntensityPointEstimate,
+    WindowedForecast,
+)
+
+
+@pytest.fixture(scope="session")
+def sample_data() -> list[CarbonIntensityPointEstimate]:
+    """Load sample carbon intensity data for testing."""
+    with open(Path(__file__).parent / "carbon_intensity_24h.csv", "r") as f:
+        csvfile = csv.reader(f, delimiter=",")
+        _ = next(csvfile)  # Skip header line
+        data = [
+            CarbonIntensityPointEstimate(
+                datetime=datetime.fromisoformat(datestr[:-1] + "+00:00"),
+                value=float(intensity_value),
+            )
+            for datestr, _, _, intensity_value in csvfile
+        ]
+        return data
+
+
+class TestParseTimeConstraint:
+    """Test the parse_time_constraint function."""
+
+    def test_parse_full_iso_datetime(self):
+        """Test parsing full ISO datetime strings."""
+        result = parse_time_constraint("2024-01-15T09:30:00")
+        expected = datetime(2024, 1, 15, 9, 30, 0)
+        assert result is not None
+        assert result.replace(tzinfo=None) == expected
+
+    def test_parse_iso_datetime_with_timezone(self):
+        """Test parsing ISO datetime with timezone info."""
+        result = parse_time_constraint("2024-01-15T09:30:00+00:00")
+        expected = datetime(2024, 1, 15, 9, 30, 0, tzinfo=timezone.utc)
+        assert result == expected
+
+    def test_parse_iso_datetime_with_z_suffix(self):
+        """Test parsing ISO datetime with Z suffix (UTC)."""
+        result = parse_time_constraint("2024-01-15T09:30:00Z")
+        expected = datetime(2024, 1, 15, 9, 30, 0, tzinfo=timezone.utc)
+        assert result == expected
+
+    def test_parse_time_only(self):
+        """Test parsing time-only strings."""
+        result = parse_time_constraint("09:30")
+        today = datetime.now().date()
+        expected_time = datetime.combine(today, time(9, 30))
+        assert result is not None
+        assert result.replace(tzinfo=None) == expected_time.replace(tzinfo=None)
+
+    def test_parse_time_with_seconds(self):
+        """Test parsing time with seconds."""
+        result = parse_time_constraint("09:30:45")
+        today = datetime.now().date()
+        expected_time = datetime.combine(today, time(9, 30, 45))
+        assert result is not None
+        assert result.replace(tzinfo=None) == expected_time.replace(tzinfo=None)
+
+    def test_parse_empty_string_returns_none(self):
+        """Test that empty string returns None."""
+        result = parse_time_constraint("")
+        assert result is None
+
+    def test_parse_none_returns_none(self):
+        """Test that None returns None."""
+        result = parse_time_constraint("")  # Pass empty string instead of None
```

Comment:
Why pass the empty string here - it turns out to be the same as the previous test?